library/lua/script-data.lua

```diff
@@ -0,0 +1,344 @@
+  else -- assume world
+    --[[
+    -- Safety check for arena mode
+    -- Arena mode saves are the only(?) worlds that can trigger world loaded/unloaded without already having an existing save file
+    -- DFhack's dfhack.getSavePath() will provide a path for that save game's folder even if it doesn't actually exist!
+
+    -- Check the save folder actually exists before attempting to save - if it doesn't, abort!
+    update_save_path()
+
+    if not dfhack.filesystem.isdir(recorded_save_path) then
+      return false
+    end
+
+    -- Otherwise, continue as usual
+    ]]
+    --^ Old safety check for before quit_without_saving was added to work around it (the check didn't really work properly anyway)
+    ensure_save_dir_exists("world")
+    path = get_save_dir("world") .. "/" .. id .. ".json"
+  end
+
+  json.encode_file(get_loaded_data(id).data, path)
+end
+
+-- Trigger saves for all persistent tables which have the autosave flag currently set
+-- This is naturally triggered by this script whenever an autosave / quicksave is detected
+function save_autosave()
+  for id, entry in pairs(loaded_data) do
+    if entry.autosave == true then
+      save(id)
+    end
+  end
+end
+
+-- Sets the autosave flag for an entry
+function set_autosave(id, do_autosave)
+  get_loaded_data(id).autosave = do_autosave
+end
+
+-- Set whether the current save should save world entries when quitting
+-- This value is reset back to false whenever a world is unloaded / loaded
+-- It's automatically set to true whenever an arena game is started
+function set_quit_without_saving(bool)
+  quit_without_saving = bool
+end
+
+-- Returns the loaded data entry (essentially the metadata of a loaded script data) for the script data of the given ID
+-- There's not likely a reason to use this externally.
+function get_loaded_data(id)
+  return loaded_data[id]
+end
+
+-- Check to see if save path info is recorded
+-- If it's not, attempt to update the record if possible
+function update_save_path()
+  if recorded_save_path == nil and dfhack.isWorldLoaded() then
+    recorded_save_path = dfhack.getSavePath()
+  end
+end
+---------------------------------------------------------------------
+-- MISC
+
+local function world_load_setup()
+  -- We store the path of the loaded save just in case there's some hiccups with the fact we try to save world data during the SC_WORLD_UNLOADED event (where dfhack.getSavePath() might possibly return nil)
+  recorded_save_path = dfhack.getSavePath()
+
+  -- Begin monitoring for autosaves / quicksaves
+  -- Since autosaves are only actually relevant in fortress mode, there's no reason to start checking for them outside of that mode
+  if dfhack.world.isFortressMode() then
+    start_loop()
+  end
+
+  -- Automatically disable saving on exit if in arena mode
+  -- This is because it's the only current mode(?) that allows for quitting without saving, and there's no way (without some hackery) to really detect that
```

Comment:
I was about to suggest checking `df.global.save_on_exit`, which should theoretically be the inverse of your `quit_without_saving`, but it looks like that is set to true in arena mode, unfortunately, now that arena mode supports saving. I suspect it's probably immediately set to false if you select "abort game" in the options menu, but I haven't confirmed. Could be worth keeping in mind in case it's useful elsewhere. See below for more.